buf/validate/internal/extra_func.cc

```diff
@@ -220,6 +220,130 @@ cel::CelValue isIP(google::protobuf::Arena* arena, cel::CelValue::StringHolder l
   return isIPvX(arena, lhs, cel::CelValue::CreateInt64(0));
 }
 
+/**
+ * IP Prefix Validation
+ */
+bool IsIpv4Prefix(const std::string_view to_validate, bool strict) {
+  std::vector<std::string> split = absl::StrSplit(to_validate.data(), '/');
+  if (split.size() != 2) {
+    return false;
+  }
+  std::string ip = split[0];
+  std::string prefixlen = split[1];
+
+  // validate ip
+  if (!IsIpv4(ip)) {
+    return false;
+  }
+
+  // validate prefixlen
+  if (prefixlen.empty()) {
+    return false;
+  }
+  int prefixlen_int;
+  if (!absl::SimpleAtoi(prefixlen, &prefixlen_int)) {
+    return false;
+  }
+  if (prefixlen_int < 0 || prefixlen_int > 32) {
+    return false;
+  }
+
+  // check host part is all-zero if strict
+  if (strict) {
+    struct sockaddr_in sa;
+    if (inet_pton(AF_INET, ip.c_str(), &sa.sin_addr) != 1) {
+      return false;
+    }
+    struct in_addr mask;
+    mask.s_addr = htonl(std::pow(2, 32 - prefixlen_int) - 1);
+    return ((sa.sin_addr.s_addr & mask.s_addr) == 0);
+  }
+
+  return true;
+}
+
+bool IsIpv6Prefix(const std::string_view to_validate, bool strict) {
+  std::vector<std::string> split = absl::StrSplit(to_validate.data(), '/');
+  if (split.size() != 2) {
+    return false;
+  }
+  std::string ip = split[0];
+  std::string prefixlen = split[1];
+
+  // validate ip
+  if (!IsIpv6(ip)) {
+    return false;
+  }
+
+  // validate prefixlen
+  if (prefixlen.empty()) {
+    return false;
+  }
+  int prefixlen_int;
+  if (!absl::SimpleAtoi(prefixlen, &prefixlen_int)) {
+    return false;
+  }
+  if (prefixlen_int < 0 || prefixlen_int > 128) {
+    return false;
+  }
+
+  // check host part is all-zero if strict
+  if (strict) {
+    struct sockaddr_in6 sa_six;
+    if (inet_pton(AF_INET6, ip.c_str(), &sa_six.sin6_addr) != 1) {
+      return false;
+    }
+    struct in6_addr mask;
+    for (int i=0; i<4; i++) {
+      if ((prefixlen_int / 32) < i) {
+        mask.__in6_u.__u6_addr32[i] = htonl(0xffffffff);
+      } else if ((prefixlen_int / 32) == i) {
+        mask.__in6_u.__u6_addr32[i] = htonl(std::pow(2, 32*(i+1) - prefixlen_int) - 1);
```

Comment:
Using `std::pow(2, 32*(i+1) - prefixlen_int)` seems a bit unnecessary. I believe you can just do `1 << (32*(i+1) - prefixlen_int)`. (It might be necessary to use `1ull` to avoid the overflow. I'm not 100% sure.)